### PR TITLE
Configure checkout session webhook events

### PR DESCRIPTION
**Pull request by Mayisha**
Towards STRIPE-921

## Changes proposed in this Pull Request:
- Add the checkout session related webhook events to the list.
    - 'checkout.session.completed'
    - 'checkout.session.expired'
    - 'checkout.session.async_payment_succeeded'
    - 'checkout.session.async_payment_failed' 
- Reconfigure webhook when adaptive pricing is enabled in the settings.
    - We are only reconfiguring if the currently registered webhook does not have the checkout session related events.

## Testing instructions
- Add the following filter to allow adaptive pricing and checkout sessions in your store-
 `add_filter('wc_stripe_is_checkout_sessions_available', '__return_true');`
- Use `develop` branch.
- As an admin, go to **WooCommerce > Settings > Payments > Stripe > Settings Tab**.
    - For a new site, connect to your Stripe account.
    - For existing store, click the `Configure connection` button and then reconfigure the webhook from the modal.
- Go to the Stripe dashboard of your connected account and check the registered webhook for your test site. Verify that the webhook has 19 events registered.
- 
Screenshot 2026-02-14 at 12 03 52 AM

- Now checkout to this branch.
- In the settings page, scroll down to the `Advanced settings` section. Enable optimized checkout and adaptive pricing. 
- Save the changes.
- Refresh the Stripe dashboard and check the registered webhook for your test site. Verify that the webhook has now 23 events registered.

### Changelog entry

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 1]
```diff
@@ -610,6 +612,10 @@ private function update_oc_settings( WP_REST_Request $request ) {
 			}
 			$current_value = $this->gateway->get_option( $attribute );
 
+			if ( 'is_ap_enabled' === $request_key && 'yes' === $value && 'yes' !== $current_value ) {
+				$adaptive_pricing_just_enabled = true;
+			}
```
Comment (daledupreez) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: There is an edge case allowed by the UI that we are not handling: as things stand, you can enable OCS and Adaptive Pricing, and then disable OCS, and then re-enable OCS, which then keeps the same value for Adaptive Pricing from before. While that's an edge case, and we should have updated the webhooks the first time around, I think we should still handle it.

File: tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php [hunk 2]
```diff
@@ -437,6 +437,77 @@ public function test_update_settings_enforces_apple_pay_google_pay_requires_card
...
 
+	/**
+	 * Tests that maybe_reconfigure_webhooks_on_update is called when Adaptive Pricing is enabled via the REST API.
+	 */
+	public function test_update_oc_settings_calls_maybe_reconfigure_webhooks_when_adaptive_pricing_enabled() {
+		$original_account = WC_Stripe::get_instance()->account;
+
+		$mock_account = $this->getMockBuilder( 'WC_Stripe_Account' )
+			->disableOriginalConstructor()
+			->setMethods( [ 'maybe_reconfigure_webhooks_on_update' ] )
```
Comment (daledupreez) [inline tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php]: Nit: `setMethods` is deprecated in PHPUnit 9. It will be more future-proof if you use `onlyMethods()` instead.
```suggestion
			->onlyMethods( [ 'maybe_reconfigure_webhooks_on_update' ] )
```

**Review by daledupreez — COMMENTED** [review]
This is working as described for the documented test cases, but we're missing a subtle edge case.

**Review by wjrosa — APPROVED** [review]
Looks good to me! Dale's comment is something we should take into account

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 1 above)
Comment (Mayisha) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: I have removed the "just enabled" check in [64d6115](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/5018/commits/64d61159ab189c2f7f421598a9deed8c5579193b) and now calling 'maybe_reconfigure_webhooks_on_update' if `is_ap_enabled` is true. In `maybe_reconfigure_webhooks_on_update`, we first check if the existing webhook is missing any required events and reconfigure the webhook only when there is any missing event.

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 3]
```diff
@@ -622,6 +628,12 @@ private function update_oc_settings( WP_REST_Request $request ) {
 				);
 			}
 		}
+
+		// If Adaptive Pricing is enabled, we may need to reconfigure the webhooks to include the checkout session events.
+		// maybe_reconfigure_webhooks_on_update only reconfigures the webhooks if the events differ from the desired events.
+		if ( $adaptive_pricing_enabled ) {
```
Comment (daledupreez) [inline includes/admin/class-wc-rest-stripe-settings-controller.php]: I have two more nits:

1. Should we only run this when both OCS and Adaptive Pricing are enabled? As I noted in my original comment, we need both to be enabled for the check to be relevant.

2. I am wondering if it would be simpler to just check the option values after the logic runs, as a caller could only specify some of the values but not all of them, in which case the `is_ap_enabled` value may not be in the request, but the option may already be set the way we want.

```suggestion
		$stripe_settings = WC_Stripe_Helper::get_settings();
		if ( 'yes' === ( $stripe_settings['optimized_checkout_element'] ?? 'no' ) && 'yes' === ( $stripe_settings['adaptive_pricing'] ?? 'no' ) ) {
```

Comment (daledupreez) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: And if we _really_ want to do this right, might we even want to do this from an action after the options get updated?

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 4]
```diff
@@ -610,6 +612,10 @@ private function update_oc_settings( WP_REST_Request $request ) {
 			}
 			$current_value = $this->gateway->get_option( $attribute );
 
+			if ( 'is_ap_enabled' === $request_key && 'yes' === $value ) {
+				$adaptive_pricing_enabled = true;
+			}
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: It was intentional to check for 'yes'.

File: includes/class-wc-stripe-account.php [hunk 5]
```diff
@@ -56,6 +56,10 @@ class WC_Stripe_Account {
 		'payment_intent.requires_action',
 		'setup_intent.succeeded',
 		'setup_intent.setup_failed',
+		'checkout.session.completed',
+		'checkout.session.expired',
+		'checkout.session.async_payment_succeeded',
+		'checkout.session.async_payment_failed',
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply includes/class-wc-stripe-account.php:62]: We will add the webhook handlers in a separate PR.

File: tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php [hunk 6]
```diff
@@ -437,6 +437,62 @@ public function test_update_settings_enforces_apple_pay_google_pay_requires_card
...
+		$request = new WP_REST_Request( 'POST', self::SETTINGS_ROUTE );
+		$request->set_param( 'is_ap_enabled', false );
+
+		$response = $this->controller->update_settings( $request );
+
+		WC_Stripe::get_instance()->account = $original_account;
+
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertSame( 'no', $this->get_gateway()->get_option( 'adaptive_pricing' ) );
+	}
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php]: Only one scenario here. So no need for dataProvider.

File: tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php [hunk 7]
```diff
@@ -437,6 +437,62 @@ public function test_update_settings_enforces_apple_pay_google_pay_requires_card
...
+
+		WC_Stripe::get_instance()->account = $original_account;
+
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertSame( 'yes', $this->get_gateway()->get_option( 'adaptive_pricing' ) );
+	}
+
+	/**
+	 * Tests that maybe_reconfigure_webhooks_on_update is called when Adaptive Pricing is enabled via the REST API.
+	 */
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php]: Typo in comment. I will fix it.

File: includes/admin/class-wc-rest-stripe-settings-controller.php [hunk 8]
```diff
@@ -622,6 +628,12 @@ private function update_oc_settings( WP_REST_Request $request ) {
 				);
 			}
 		}
+
+		// If Adaptive Pricing is enabled, we may need to reconfigure the webhooks to include the checkout session events.
+		// maybe_reconfigure_webhooks_on_update only reconfigures the webhooks if the events differ from the desired events.
+		if ( $adaptive_pricing_enabled ) {
+			WC_Stripe::get_instance()->account->maybe_reconfigure_webhooks_on_update();
+		}
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: Not relevant after the latest commits.

File: tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php [hunk 9]
```diff
@@ -437,6 +437,77 @@ public function test_update_settings_enforces_apple_pay_google_pay_requires_card
...
+	 *
+	 * @return array[]
+	 */
+	public function provide_adaptive_pricing_webhook_not_called() {
+		return [
+			'AP disabled in request'     => [ false, 'yes' ],
+			'AP unchanged (already on)'  => [ true, 'yes' ],
+			'AP unchanged (already off)' => [ false, 'no' ],
+		];
+	}
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply tests/phpunit/Admin/WC_REST_Stripe_Settings_Controller_Test.php]: Not relevant after the latest commits.

File: includes/class-wc-stripe-account.php [hunk 10]
```diff
@@ -447,7 +453,35 @@ public function get_existing_webhook() {
...
+		$was_oc_enabled = 'yes' === ( $old_value['optimized_checkout_element'] ?? '' );
+		$was_ap_enabled = 'yes' === ( $old_value['adaptive_pricing'] ?? '' );
+
+		// If Adaptive Pricing and Optimized Checkout are both unchanged in the new value, do nothing.
+		if ( $was_ap_enabled === $is_ap_enabled && $was_oc_enabled === $is_oc_enabled ) {
+			return;
+		}
+
+		$this->maybe_reconfigure_webhooks_on_update();
+	}
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply includes/class-wc-stripe-account.php]: Added.

File: includes/class-wc-stripe-account.php [hunk 11]
```diff
@@ -447,7 +453,35 @@ public function get_existing_webhook() {
...
+			return;
+		}
+
+		$was_oc_enabled = 'yes' === ( $old_value['optimized_checkout_element'] ?? '' );
+		$was_ap_enabled = 'yes' === ( $old_value['adaptive_pricing'] ?? '' );
+
+		// If Adaptive Pricing and Optimized Checkout are both unchanged in the new value, do nothing.
+		if ( $was_ap_enabled === $is_ap_enabled && $was_oc_enabled === $is_oc_enabled ) {
+			return;
+		}
```
Comment (daledupreez) [inline includes/class-wc-stripe-account.php]: As noted by CodeRabbit, I think we need to make this code more resilient against unexpected data in both $old_value and $value.

```suggestion
		if ( ! is_array( $value ) ) {
			return;
		}
		$is_oc_enabled = 'yes' === ( $value['optimized_checkout_element'] ?? '' );
		$is_ap_enabled = 'yes' === ( $value['adaptive_pricing'] ?? '' );

		// If Adaptive Pricing or Optimized Checkout is disabled in the new value, do nothing.
		if ( ! $is_ap_enabled || ! $is_oc_enabled ) {
			return;
		}

		if ( is_array( $old_value ) ) {
			$was_oc_enabled = 'yes' === ( $old_value['optimized_checkout_element'] ?? '' );
			$was_ap_enabled = 'yes' === ( $old_value['adaptive_pricing'] ?? '' );
		} else {
			$was_oc_enabled = false;
			$was_ap_enabled = false;
		}

		// If Adaptive Pricing and Optimized Checkout are both unchanged in the new value, do nothing.
		if ( $was_ap_enabled === $is_ap_enabled && $was_oc_enabled === $is_oc_enabled ) {
			return;
		}
```

File: tests/phpunit/WC_Stripe_Account_Test.php [hunk 12]
```diff
@@ -527,4 +527,113 @@ public function test_reconfigure_webhooks_on_update_with_current_webhooks() {
...
+
+	/**
+	 * Data provider for maybe_reconfigure_webhooks_after_adaptive_pricing_enabled tests.
+	 * Covers the update_option_woocommerce_stripe_settings hook behavior: reconfigure only when
+	 * both AP and OC are enabled in the new value and at least one of them changed from the old value.
+	 *
+	 * @return array[] Old value, new value, whether reconfigure is expected, and scenario name.
+	 */
+	public function provider_maybe_reconfigure_webhooks_after_adaptive_pricing_enabled() {
+		return [
```
Comment (daledupreez) [inline tests/phpunit/WC_Stripe_Account_Test.php]: We should add some other values like null and undefined to the tests, as well as adding test cases for `$old_value === false`.

File: includes/class-wc-stripe.php [hunk 13]
```diff
@@ -648,6 +651,42 @@ public function gateway_settings_update( $settings, $old_settings ) {
...
+			return;
+		}
+
+		$was_oc_enabled = is_array( $old_value ) && 'yes' === ( $old_value['optimized_checkout_element'] ?? '' );
+		$was_ap_enabled = is_array( $old_value ) && 'yes' === ( $old_value['adaptive_pricing'] ?? '' );
+
+		// If Adaptive Pricing and Optimized Checkout are both unchanged in the new value, do nothing.
+		if ( $was_ap_enabled === $is_ap_enabled && $was_oc_enabled === $is_oc_enabled ) {
+			return;
+		}
```
Comment (daledupreez) [inline includes/class-wc-stripe.php]: 1000% non-blocking: we know that `$is_ap_enabled` and `$is_oc_enabled` must both be true, so we can focus the checks on whether the old values were false.
```suggestion
		// If Adaptive Pricing and Optimized Checkout were both enabled before, do nothing.
		if ( $was_ap_enabled && $was_oc_enabled ) {
			return;
		}
```

File: tests/phpunit/WC_Stripe_Test.php [hunk 14]
```diff
@@ -545,4 +545,120 @@ public function provide_test_add_gateways(): array {
...
+					'optimized_checkout_element' => 'yes',
+				],
+				[
+					'adaptive_pricing'           => 'no',
+					'optimized_checkout_element' => 'no',
+				],
+				false,
+			],
+		];
+	}
```
[the remark this answers is not in the document]
Comment (daledupreez) [reply tests/phpunit/WC_Stripe_Test.php]: This is what I was going to say!

File: tests/phpunit/WC_Stripe_Test.php [hunk 15]
```diff
@@ -545,4 +545,120 @@ public function provide_test_add_gateways(): array {
...
+				[
+					'adaptive_pricing'           => 'no',
+					'optimized_checkout_element' => 'yes',
+				],
+				[
+					'adaptive_pricing'           => 'yes',
+					'optimized_checkout_element' => 'yes',
+				],
+				true,
+			],
```
Comment (daledupreez) [inline tests/phpunit/WC_Stripe_Test.php]: With all the vertical space, I am finding it hard to track which values are in which argument. It might be easier to use keys to make this a bit more readable.
```suggestion
			'AP and OC newly enabled'              => [
				'old_value'   => [
					'adaptive_pricing'           => 'no',
					'optimized_checkout_element' => 'no',
				],
				'new_value'.  => [
					'adaptive_pricing'           => 'yes',
					'optimized_checkout_element' => 'yes',
				],
				'expect_call' => true,
			],
			'AP newly enabled, OC already enabled' => [
				'old_value'   => [
					'adaptive_pricing'           => 'no',
					'optimized_checkout_element' => 'yes',
				],
				'new_value'   => [
					'adaptive_pricing'           => 'yes',
					'optimized_checkout_element' => 'yes',
				],
				'expect_call' => true,
			],
```

**Review by daledupreez — APPROVED** [review]
I have some minor suggestions, but they are pretty minor. Thanks for sticking with this PR.

File: includes/admin/class-wc-rest-stripe-settings-controller.php (hunk 3 above)
Comment (Mayisha) [reply includes/admin/class-wc-rest-stripe-settings-controller.php]: > might we even want to do this from an action after the options get updated?

Thanks for the suggestion Dale ✌️

File: tests/phpunit/WC_Stripe_Test.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mayisha) [reply tests/phpunit/WC_Stripe_Test.php]: Added tests for these scenarios.

File: includes/class-wc-stripe.php (hunk 13 above)
Comment (Mayisha) [reply includes/class-wc-stripe.php]: Makes sense 👍

File: includes/class-wc-stripe-account.php (hunk 11 above)
Comment (Mayisha) [reply includes/class-wc-stripe-account.php]: Done.
